util/proactor_base.cc

```diff
@@ -82,53 +84,63 @@ void ProactorBase::Stop() {
   VLOG(1) << "Proactor::StopFinish";
 }
 
-uint32_t ProactorBase::AddIdleTask(IdleTask f) {
+uint32_t ProactorBase::AddOnIdleTask(OnIdleTask f) {
   DCHECK(InMyThread());
 
-  auto id = next_task_id_++;
-  auto res = on_idle_map_.emplace(id, std::move(f));
-  CHECK(res.second);
-  idle_it_ = on_idle_map_.begin();  // reset position to the first item.
+  uint32_t res = on_idle_arr_.size();
+  on_idle_arr_.push_back(OnIdleWrapper{.task = std::move(f), .next_ts = 0});
 
-  return id;
+  return res;
 }
 
-void ProactorBase::RunOnIdleTasks() {
-  if (on_idle_map_.empty())
-    return;
-
-  if (idle_it_ == on_idle_map_.end()) {
-    idle_it_ = on_idle_map_.begin();
-  }
+bool ProactorBase::RunOnIdleTasks() {
+  if (on_idle_arr_.empty())
+    return false;
 
   uint64_t start = GetClockNanos();
-  tl_info_.monotonic_time = start;
+  uint64_t curr_ts = start;
+
+  bool should_spin = false;
 
-  // Perform round robin with idle_it_ saving the position between runs.
+  DCHECK_LT(on_idle_next_, on_idle_arr_.size());
+
+  // Perform round robin with on_idle_next_ saving the position between runs.
   do {
-    bool res = idle_it_->second();
+    OnIdleWrapper& on_idle = on_idle_arr_[on_idle_next_];
 
-    if (!res) {
-      on_idle_map_.erase(idle_it_);
-      idle_it_ = on_idle_map_.begin();
-      break;
+    if (on_idle.task && on_idle.next_ts <= curr_ts) {
+      tl_info_.monotonic_time = curr_ts;
+
+      uint32_t level = on_idle.task();  // run the task
+
+      curr_ts = GetClockNanos();
```

Comment:
If `GetClockNanos` is non monotonic, then you might end up jumping back in time and even running the same task twice (if it has a high level and a small delta). Or is that highly unlikely? So a `curr_ts = max(curr_ts, GetClockNanos())` would solve it but probably bring other problems. Maybe its even not important at all

Reply:
yeah, this is not what worries me. as long as we won't have weird integer overflow i do not care.
the corner case is when someone changes the system clock. we will need to test it some day anyway.